projects/core/src/occ/utils/occ-fields.spec.ts

```diff
@@ -0,0 +1,193 @@
+    ).toEqual('ala(lubi(koty,i,psy),chodzi,pieszo),ula,ma');
+  });
+
+  it('should optimize BASIC, DEFAULT and FULL fields', () => {
+    expect(mergeFields(['test(FULL)', 'test(DEFAULT)', 'test(BASIC)'])).toEqual(
+      'test(FULL)'
+    );
+
+    expect(mergeFields(['test(FULL)', 'test(ala,ma)'])).toEqual(
+      'test(FULL,ala,ma)'
+    );
+  });
+
+  it('should accept string and AST objects for fields definition', () => {
+    expect(mergeFields(['ala', { ma: { kota: {} } }])).toEqual('ala,ma(kota)');
+  });
+});
+
+describe('optimizeFields', () => {
+  it('should remove BASIC and DEFAULT if FULL is present', () => {
+    expect(
+      optimizeFields({ BASIC: {}, FULL: {}, DEFAULT: {}, test: {} })
+    ).toEqual({
+      FULL: {},
+      test: {},
+    });
+  });
+  it('should remove BASIC if DEFAULT is present', () => {
+    expect(optimizeFields({ BASIC: {}, DEFAULT: {} })).toEqual({
+      DEFAULT: {},
+    });
+  });
+  it('should optimize nested fields', () => {
+    expect(optimizeFields({ ala: { DEFAULT: {}, BASIC: {} } })).toEqual({
+      ala: { DEFAULT: {} },
+    });
+  });
+});
+
+describe('parseFields', () => {
+  it('should parse basic fields', () => {
+    expect(parseFields('ala,ma,kota')).toEqual({ ala: {}, ma: {}, kota: {} });
+  });
+  it('should parse nested fields', () => {
+    expect(parseFields('ala(lubi(koty)),ula')).toEqual({
+      ala: {
+        lubi: {
+          koty: {},
+        },
+      },
+      ula: {},
+    });
+
+    expect(parseFields('ala(lubi(i,psy),ula)')).toEqual({
+      ala: {
+        lubi: {
+          i: {},
+          psy: {},
+        },
+        ula: {},
+      },
+    });
+  });
+  it('should parse complex fields', () => {
+    expect(
+      parseFields(
+        'ala,nie,wie(albo,i(co),wie(ale,nie,wiadomo)),ze,ona(lubi(koty)),ula'
+      )
+    ).toEqual({
+      ala: {},
+      nie: {},
+      wie: {
+        albo: {},
+        i: {
+          co: {},
+        },
+        wie: {
+          ale: {},
+          nie: {},
+          wiadomo: {},
+        },
+      },
+      ze: {},
+      ona: {
+        lubi: {
+          koty: {},
+        },
+      },
+      ula: {},
+    });
+  });
+  it('should parse not strictly valid fields', () => {
+    expect(parseFields('ala(lubi(koty())),,,ula')).toEqual({
+      ala: {
+        lubi: {
+          koty: {},
+        },
+      },
+      ula: {},
+    });
+  });
+});
+
+describe('stringifyFields', () => {
+  it('should stringify simple fields', () => {
+    expect(stringifyFields({ ala: {}, ma: {}, kota: {} })).toEqual(
+      'ala,ma,kota'
+    );
+  });
+  it('should stringify nested fields', () => {
+    expect(
+      stringifyFields({
+        ala: {},
+        ma: { o: {} },
+        kota: { i: { psa: {}, burka: {} } },
+      })
+    ).toEqual('ala,ma(o),kota(i(psa,burka))');
+  });
+});
+
+describe('extractFields for given fields definition', () => {
+  const data: any = {
```

Comment:
For tests safety, can we create the `data` in `beforeEach` in case the method could modify the object.

Reply:
👍